decksite/data/rule.py

```diff
@@ -0,0 +1,189 @@
+from typing import List
+
+from decksite.data import deck
+from decksite.database import db
+from magic.models import Deck
+from shared.container import Container
+
+
+def num_classified_decks() -> int:
+    sql = 'SELECT COUNT(DISTINCT(deck_id)) AS c FROM ({apply_rules_query}) AS applied_rules'.format(apply_rules_query=apply_rules_query())
+    return (db().select(sql))[0]['c']
+
+def mistagged_decks() -> List[Deck]:
+    sql = """SELECT
+                    deck_id, rule_archetype.name AS rule_archetype_name, tagged_archetype.name AS tagged_archetype_name
+                FROM
+                    ({apply_rules_query}) AS applied_rules
+                JOIN
+                    deck
+                ON
+                    applied_rules.deck_id = deck.id
+                JOIN
+                    archetype AS rule_archetype
+                ON
+                    rule_archetype.id = applied_rules.archetype_id
+                JOIN
+                    archetype AS tagged_archetype
+                ON
+                    tagged_archetype.id = deck.archetype_id
+                WHERE
+                    rule_archetype.id != tagged_archetype.id""".format(apply_rules_query=apply_rules_query())
+    deck_ids: List[str] = []
+    rule_archetypes = {}
+    for r in (Container(row) for row in db().select(sql)):
+        deck_ids.append(str(r.deck_id))
+        rule_archetypes[r.deck_id] = r.rule_archetype_name
+    if len(deck_ids) == 0:
+        return []
+    result = deck.load_decks(where='d.id IN ('+','.join(deck_ids)+')')
+    for d in result:
+        d.rule_archetype_name = rule_archetypes[d.id]
+    return result
+
+def doubled_decks() -> List[Deck]:
+    sql = """SELECT
+                    deck_id, GROUP_CONCAT(archetype.name) AS concat_archetypes
```

Comment:
Strictly speaking in the super spaced out SQL style each column selected gets its own line.